### PR TITLE
pytorch-forecasting 0.10.3 :snowflake:

### PR DESCRIPTION
**Pull request by skupr-anaconda**
Changelog: https://github.com/jdb78/pytorch-forecasting/blob/v0.10.3/CHANGELOG.md
License: https://github.com/jdb78/pytorch-forecasting/blob/v0.10.3/LICENSE
Requirements: https://github.com/jdb78/pytorch-forecasting/blob/v0.10.3/pyproject.toml

Actions:
1. Clean up the recipe: 
  - Remove `noarch python`
  - Skip `py<38` and `py>310`
  - Add flags `--no-deps --no-build-isolation` to `script`
  - Fix `python` in `host` and `run`
2. Add a patch to fix:
  - `"RuntimeError: This does not appear to be a Git project"`. The issue was related to `pip` but I can't resolve it like here https://github.com/mtkennerly/poetry-dynamic-versioning/issues/46 or here https://github.com/mtkennerly/poetry-dynamic-versioning/issues/90
  - compatibility with **optuna 3.x.x**. See the comment below https://github.com/AnacondaRecipes/pytorch-forecasting-feedstock/pull/2#issuecomment-1632496772
4. Update runtime dependencies and pinnings.
5. Constrain `numpy` to **<1.24** in `test/requires` to prevent `AttributeError: module 'numpy' has no attribute 'float'.`
6. Fix intel-openmp 2023.1 and llvm-openmp 14.0.6 compatibility on osx-64 leading to `Segmentation fault: 11.` error for a lot of tests. 
7. Disable `pip check` on `win64` because of `tensorboard 2.10.0 has requirement protobuf<3.20,>=3.9.2, but you have protobuf 3.20.3.` but it looks like the package works correctly 
8. Disable `pip check` on `linux-ppc64le` because of `tensorboard 2.6.0 has requirement google-auth<2,>=1.6.3, but you have google-auth 2.6.0`.
9. Skip tests for `win` and `linux and ppc64le` (see comments in the recipe)

### CODE REVIEW

**Comment by skupr-anaconda** [comment]
I patched the version of `optuna` dependency to add a support for `>=3.0.0`

`optuna` is used only in one file `tuning.py` https://github.com/search?q=repo%3Ajdb78%2Fpytorch-forecasting%20optuna&type=code
If you make a `diff` of the files [tuning.py of v1.0.0](https://raw.githubusercontent.com/jdb78/pytorch-forecasting/v1.0.0/pytorch_forecasting/models/temporal_fusion_transformer/tuning.py) and [tuning.py of v0.10.3](https://raw.githubusercontent.com/jdb78/pytorch-forecasting/v0.10.3/pytorch_forecasting/models/temporal_fusion_transformer/tuning.py) we will see that code related to `optuna` hasn't been changed significantly (only remove an import of `TensorBoardCallback`):
```
╰─❯ diff --color='always' -U1 tuning_v1.0.0.py tuning_v0.10.3.py                    
--- tuning_v1.0.0.py   2023-07-12 15:57:27.754739221 +0300
+++ tuning_v0.10.3.py    2023-07-12 15:57:52.021056136 +0300
@@ -8,10 +8,10 @@
 
-import lightning.pytorch as pl
-from lightning.pytorch.callbacks import LearningRateMonitor, ModelCheckpoint
-from lightning.pytorch.loggers import TensorBoardLogger
-from lightning.pytorch.tuner import Tuner
 import numpy as np
 import optuna
-from optuna.integration import PyTorchLightningPruningCallback
+from optuna.integration import PyTorchLightningPruningCallback, TensorBoardCallback
 import optuna.logging
+import pytorch_lightning as pl
+from pytorch_lightning import Callback
+from pytorch_lightning.callbacks import LearningRateMonitor
+from pytorch_lightning.loggers import TensorBoardLogger
 import statsmodels.api as sm
@@ -27,5 +27,11 @@
 
-# need to inherit from callback for this to work
-class PyTorchLightningPruningCallbackAdjusted(pl.Callback, PyTorchLightningPruningCallback):
-    pass
+class MetricsCallback(Callback):
+    """PyTorch Lightning metric callback."""
+
+    def __init__(self):
+        super().__init__()
+        self.metrics = []
+
+    def on_validation_end(self, trainer, pl_module):
+        self.metrics.append(trainer.callback_metrics)
 
@@ -117,3 +123,3 @@
         # Filenames for each trial must be made unique in order to access each checkpoint.
-        checkpoint_callback = ModelCheckpoint(
+        checkpoint_callback = pl.callbacks.ModelCheckpoint(
             dirpath=os.path.join(model_path, "trial_{}".format(trial.number)), filename="{epoch}", monitor="val_loss"
@@ -121,2 +127,6 @@
 
+        # The default logger in PyTorch Lightning writes to event files to be consumed by
+        # TensorBoard. We don't use any logger here as it requires us to implement several abstract
+        # methods. Instead we setup a simple callback, that saves metrics from each validation step.
+        metrics_callback = MetricsCallback()
         learning_rate_callback = LearningRateMonitor()
@@ -125,3 +135,3 @@
         default_trainer_kwargs = dict(
-            accelerator="auto",
+            gpus=[0] if torch.cuda.is_available() else None,
             max_epochs=max_epochs,
@@ -129,5 +139,6 @@
             callbacks=[
+                metrics_callback,
                 learning_rate_callback,
                 checkpoint_callback,
-                PyTorchLightningPruningCallbackAdjusted(trial, monitor="val_loss"),
+                PyTorchLightningPruningCallback(trial, monitor="val_loss"),
             ],
@@ -163,3 +174,3 @@
                 gradient_clip_val=gradient_clip_val,
-                accelerator="auto",
+                gpus=[0] if torch.cuda.is_available() else None,
                 logger=False,
@@ -168,4 +179,3 @@
             )
-            tuner = Tuner(lr_trainer)
-            res = tuner.lr_find(
+            res = lr_trainer.tuner.lr_find(
                 model,
@@ -201,3 +211,3 @@
         # report result
-        return trainer.callback_metrics["val_loss"].item()
+        return metrics_callback.metrics[-1]["val_loss"].item()
```

**Review by varlackc — COMMENTED** [review]
we might want to remove the `abs.yaml` file before merging to defaults.

  Other than that LGMT

File: recipe/meta.yaml
Comment (akabanovs) [inline recipe/meta.yaml]: Instead of this, you could try adding 

- blas=*=openblas  # [osx and x86_64]

to host and run sections. This basically forces to pick scipy (and any other suitable packages in the chain, like numpy) that was built with openblas rather than mkl. Openblas typically pairs with llvm-openmp and mkl with intel-openmp on osx-64. By default, mkl variants are picked. By picking the variant that uses openblas rather than mkl, intel-openmp will not be pulled in and so you will not have llvm-openmp / intel-openmp conflict.  Sumit said this whole mess is coming from mkl variant of numpy as it tends to import both llvm-openmp and intel-openmp, but these 2 should not really be used together.

File: recipe/meta.yaml
Comment (akabanovs) [inline recipe/meta.yaml:56]: Possibly add here too:

blas=*=openblas # [osx and x86_64]

to pick openblas (and hence no intel-openmp) variant of numpy.

File: recipe/meta.yaml
Comment (akabanovs) [inline recipe/meta.yaml]: Interesting. I thought Openblas/MKL include implementation of LAPACK.

File: recipe/meta.yaml
Comment (akabanovs) [inline recipe/meta.yaml]: Perhaps won't be needed. See comment beneath.

**Comment by akabanovs** [comment]
I am off on holiday for 2 weeks and will not be able to re-review. Please see my comments above, but generally looks good.

**Review by varlackc — APPROVED** [review]
LGTM

File: recipe/meta.yaml
Comment (skupr-anaconda) [reply recipe/meta.yaml]: Thanks. That's a great explanation!

File: recipe/meta.yaml
Comment (skupr-anaconda) [reply recipe/meta.yaml:56]: Added

File: recipe/meta.yaml
Comment (boldorider4) [inline recipe/meta.yaml]: 1.5.2 is available

Comment (skupr-anaconda) [reply recipe/meta.yaml]: I'm using the lowest available for python versions

Comment (boldorider4) [reply recipe/meta.yaml]: isn't 1.5.2 also available for all py versions on all platforms? Do we still have to use a lower one?

Comment (skupr-anaconda) [reply recipe/meta.yaml]: I'm using this logic only to be sure that we can build the package with the lowest available version of the python build tool (poetry-core).

File: recipe/meta.yaml
Comment (M-Waszkiewicz-Anaconda) [inline recipe/meta.yaml:30]: don't we need setuptools and wheel ?

File: recipe/meta.yaml
Comment (boldorider4) [reply recipe/meta.yaml]: I thought we wanted to use the highest available version in this case. I may be wrong.__

**Review by boldorider4 — APPROVED** [review]

File: recipe/fix_versions.patch
Comment (rafaelmartins) [inline recipe/fix_versions.patch]: this is the type of issue that is better fixed upstream. Why is dynamic version not working?

File: recipe/meta.yaml
Comment (rafaelmartins) [reply recipe/meta.yaml]: using the lowest available version would be relevant if this was a source distro, as it would provide an upgrade path for people with old installations. for a binary distribution it is not a problem, it should be fine to use the latest supported version.

File: recipe/meta.yaml
Comment (rafaelmartins) [inline recipe/meta.yaml]: I think that just adding an upper limit to optana in pyproject.toml patch is not enough if you don't enforce it here too. This may result in some repodata hotfix when new optana is released, which we may want to avoid.

File: recipe/meta.yaml
Comment (rafaelmartins) [inline recipe/meta.yaml]: if you keep the pyproject.toml patch you can remove this

File: recipe/meta.yaml
Comment (rafaelmartins) [inline recipe/meta.yaml]: I think that your dynamic versioning problems are happening because of this. Some specific reason to use a git tag tarball instead of the pypi release?

https://pypi.org/project/pytorch-forecasting/0.10.3/#files

File: recipe/meta.yaml
Comment (rafaelmartins) [inline recipe/meta.yaml]: why?

**Review by rafaelmartins — CHANGES_REQUESTED** [review]

File: recipe/meta.yaml
Comment (skupr-anaconda) [reply recipe/meta.yaml]: Thanks for pointing to this URL. I replaced PyPi url to get the `tests` folder ([pypi](https://files.pythonhosted.org/packages/02/f0/92eb7e55f665de688120590bcee95b6704f8c6c07ae07d1255ecf734997e/pytorch_forecasting-0.10.3.tar.gz) hasn't it) for further testing the package. And without testing, I couldn't find all issues

File: recipe/meta.yaml
Comment (skupr-anaconda) [reply recipe/meta.yaml]: Thanks. I'll use a plain `poetry-core`

File: recipe/meta.yaml
Comment (skupr-anaconda) [reply recipe/meta.yaml]: This package is still in build-system.requires of the pyproject.toml so the build will fail without it

File: recipe/meta.yaml
Comment (skupr-anaconda) [reply recipe/meta.yaml]: Thanks. Then I'll use it with the upper bound - `optuna >=2.3.0,<4.0.0`

File: recipe/meta.yaml
Comment (skupr-anaconda) [reply recipe/meta.yaml]: I left comments about tensorboard above. If we skip platfroms by a selector then `anaconda-linter` thinks that `pip check` is missing for all platforms. I think it's an anaconda-linter bug

File: recipe/meta.yaml
Comment (skupr-anaconda) [reply recipe/meta.yaml:30]: No, if poetry-core is used then they are redandunt

File: recipe/fix_versions.patch
Comment (skupr-anaconda) [reply recipe/fix_versions.patch]: dynamic version is not working because of `RuntimeError: This does not appear to be a Git project.`

File: recipe/meta.yaml
Comment (rafaelmartins) [reply recipe/meta.yaml]: this should be discussed upstream, probably sending a PR to include tests in sdist. github tag tarball misses PKG-INFO, that is probably why dynamic versioning fails. I'd say fetch both files and copy just the tests from github tag tarball until that's fixed upstream. it is probably better than patching pyproject.toml

File: recipe/meta.yaml
Comment (rafaelmartins) [reply recipe/meta.yaml]: I see, but it is not because the tests are not run that `pip` and `wheel` dependencies can be skipped, as they are not strictly test-related. even if you want to skip the checks, I see no reason to silent the wheel lint.

File: recipe/meta.yaml
Comment (rafaelmartins) [reply recipe/meta.yaml]: then the patch is incomplete. if you are not using poetry-dynamic-versioning as build tool anywhere, there's no reason to have it in build-system.requires.

File: recipe/meta.yaml
Comment (skupr-anaconda) [reply recipe/meta.yaml]: dynamic versioning fails even if I use PyPi url.

File: recipe/meta.yaml
Comment (skupr-anaconda) [reply recipe/meta.yaml]: Ah, I see. I'll remove it

File: recipe/meta.yaml
Comment (rafaelmartins) [reply recipe/meta.yaml]: this is wrong. the sdist file won't work as is, even with PKG-INFO file included. The source tarballs from upstream are seriously broken. I'd suggest discussing it with maintainer, and applying the same patches here until they are released upstream, instead of carrying custom patches.

Comment (rafaelmartins) [reply recipe/meta.yaml]: > github tag tarball misses PKG-INFO, that is probably why dynamic versioning fails.

this is wrong. the sdist file won't work as is, even with PKG-INFO file included. The source tarballs from upstream are seriously broken. I'd suggest discussing it with maintainer, and applying the same patches here until they are released upstream, instead of carrying custom patches.

Comment (rafaelmartins) [reply recipe/meta.yaml]: > dynamic versioning fails even if I use PyPi url.

yes, I found that out just now. this must be discussed further with upstream

File: recipe/meta.yaml
Comment (skupr-anaconda) [reply recipe/meta.yaml]: I'll rewrite it

Comment (skupr-anaconda) [reply recipe/meta.yaml]: Done

File: recipe/meta.yaml
Comment (skupr-anaconda) [reply recipe/meta.yaml]: I've opened the issue https://github.com/jdb78/pytorch-forecasting/issues/1350

Comment (skupr-anaconda) [reply recipe/meta.yaml]: @rafaelmartins Do you think this is blocking us to deliver the package to the destination channel? Because I think if we can build and test the package that it's working we can go ahead

Comment (rafaelmartins) [reply recipe/meta.yaml]: there's a cleaner fix for this issue, without waiting on upstream. you can create a file called `.git_archival.json`, in the same folder as `pyproject.toml`, with the following content:

```json
{
  "hash-full": "7f67513a8e9a7985de831b3269d4f7692cdf3380",
  "hash-short": "7f67513",
  "timestamp": "2022-09-05T13:00:35+01:00",
  "refs": "HEAD, tag: v0.10.3",
  "describe": "v0.10.3"
}
```

Reference: https://github.com/mtkennerly/dunamai#vcs-archives

(Edit: the hashes were pointing to 0.10.2, now it is correct)

Comment (rafaelmartins) [reply recipe/meta.yaml]: This way the poetry plugin will detect the version without the need of any patching. This is the file that is missing in the sdist tarball

Comment (skupr-anaconda) [reply recipe/meta.yaml]: Thanks. That's much better, and it's great to learn something new

File: recipe/meta.yaml
Comment (rafaelmartins) [inline recipe/meta.yaml:92]: does adding wheel to the dependencies cause any problem?

**Review by rafaelmartins — APPROVED** [review]

File: recipe/meta.yaml
Comment (skupr-anaconda) [reply recipe/meta.yaml:92]: No, but @JeanChristopheMorinPerso has already pointed https://github.com/AnacondaRecipes/prince-feedstock/pull/1#discussion_r1244001638 that it's redundant for `poetry-core`

Comment (rafaelmartins) [reply recipe/meta.yaml:92]: cool, I'll send some PRs to the linter when I find some time, to fix the false positives for `pip check` when it is explicitly disabled, this wheel redundancy and maybe the blas dependency declarations

**Review by M-Waszkiewicz-Anaconda — APPROVED** [review]
LGTM

**Pull request merged**
